File: python/21-MergeTwoSortedLists/main.py
from typing import Optional

import sys
sys.path.append('../modules')
from leetcode import ListNode

import unittest


class Solution:
    def mergeTwoLists(self, list1: Optional[ListNode], list2: Optional[ListNode]) -> Optional[ListNode]:
        if not list1:
            return list2
        
        if not list2:
            return list1
        
        if list1.val < list2.val:
            return ListNode(list1.val, self.mergeTwoLists(list1.next, list2))
        else:
            return ListNode(list2.val, self.mergeTwoLists(list1, list2.next))

class Tests(unittest.TestCase):
    def test_example1(self):
        solution = Solution()
        
        s = solution.mergeTwoLists(ListNode.makeList(1, 2, 4), ListNode.makeList(1, 3, 4))
        values = ListNode.flatten(s)

        self.assertListEqual(values, [1, 1, 2, 3, 4, 4])

if __name__ == '__main__':
    unittest.main()

File: elixir/21-MergeTwoSortedLists/main.exs
defmodule ListNode do
  @type t :: %__MODULE__{
          val: integer,
          next: ListNode.t() | nil
        }
  defstruct val: 0, next: nil
end

defmodule Solution do
  @spec merge_two_lists(list1 :: ListNode.t | nil, list2 :: ListNode.t | nil) :: ListNode.t | nil
  def merge_two_lists(list1, list2) when list1.val < list2.val  do
    %ListNode{val: list1.val, next: merge_two_lists(list1.next, list2)}
  end

  def merge_two_lists(list1, list2) when list1.val >= list2.val do
    %ListNode{val: list2.val, next: merge_two_lists(list1, list2.next)}
  end

  def merge_two_lists(list1, nil) do
    list1
  end

  def merge_two_lists(nil, list2) do
    list2
  end
end

defmodule Tests do
  defp list_to_string(list) when list == nil do
    ""
  end

  defp list_to_string(list) do
    "#{list.val} " <> list_to_string(list.next)
  end

  def example1 do
    list1 = %ListNode{val: 1, next: %ListNode{val: 2, next: %ListNode{val: 4}}}
    list2 = %ListNode{val: 1, next: %ListNode{val: 3, next: %ListNode{val: 4}}}
    IO.puts(list_to_string(Solution.merge_two_lists(list1, list2)))
  end

  def example2 do
    IO.puts(list_to_string(Solution.merge_two_lists(nil, nil)))
  end

  def example3 do
    IO.puts(list_to_string(Solution.merge_two_lists(nil, %ListNode{val: 0})))
  end
end

Tests.example1
Tests.example2
Tests.example3